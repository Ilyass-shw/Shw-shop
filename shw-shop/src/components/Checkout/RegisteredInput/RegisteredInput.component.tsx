import styled from 'styled-components';
import { FieldError } from 'react-hook-form';

type InputTypeProps = {
  isError: FieldError | undefined;
};
export const Input = styled.input<InputTypeProps>`
  border-radius: 5px;
  border: 1px solid ${({ isError }) => (isError ? 'red' : '#d9d9d9')};
  padding: 1.2rem;
  margin: 0.9rem 0;
  font-size: 0.9rem;
  &:focus-visible {
    border-color: red;
  }
`;

// ============= ============= ============= ============= =============

export const FormError = styled.p`
  color: red;
  margin: 0 0 0.8rem 1rem;
`;
