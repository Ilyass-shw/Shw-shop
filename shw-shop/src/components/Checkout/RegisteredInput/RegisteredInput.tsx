import React from 'react';
import { useFormContext, RegisterOptions } from 'react-hook-form';

import { Input, FormError } from './RegisteredInput.component';

interface RegisteredInput extends React.InputHTMLAttributes<HTMLInputElement> {
  Name: string;
  registerOptions?: RegisterOptions;
}

const RegisteredInput: React.FC<RegisteredInput> = (props) => {
  const { Name, registerOptions, ...InputProps } = props;
  const {
    register,
    formState: { errors },
  } = useFormContext();
  return (
    <>
      <Input
        {...register(Name, registerOptions)}
        isError={errors[Name]}
        {...InputProps}
      />
      <FormError>{errors[Name] && errors[Name].message}</FormError>
    </>
  );
};

export default RegisteredInput;
