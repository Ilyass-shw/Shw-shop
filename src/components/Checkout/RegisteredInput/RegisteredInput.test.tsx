import React from 'react';
import userEvent from '@testing-library/user-event';
import {
  renderWithReactHookForm,
  screen,
  waitFor,
} from '../../../test-utils/testUtils';
import RegisteredInput from './RegisteredInput';

describe('RegisteredInput', () => {
  const saveData = jest.fn();

  it('should render', () => {
    renderWithReactHookForm(
      <RegisteredInput
        Name="email"
        registerOptions={{ required: 'This is required.' }}
        type="email"
        placeholder="Email"
        autoCapitalize="off"
        spellCheck="false"
        autoComplete="email"
      />,
      saveData,
    );
  });

  it('should match snapshot', () => {
    const { asFragment } = renderWithReactHookForm(
      <RegisteredInput
        Name="email"
        registerOptions={{ required: 'This is required.' }}
        type="email"
        placeholder="Email"
        autoCapitalize="off"
        spellCheck="false"
        autoComplete="email"
      />,
      saveData,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should watch input correctly', () => {
    renderWithReactHookForm(
      <RegisteredInput Name="phone" type="tel" placeholder="Phone " />,
      saveData,
    );
    const registeredInput = screen.getByPlaceholderText('Phone');
    userEvent.type(registeredInput, 'example.net@gmail.com');
    expect(registeredInput).toHaveValue('example.net@gmail.com');
  });

  it('should display correct error message and submit correctly', async () => {
    renderWithReactHookForm(
      <RegisteredInput
        Name="email"
        type="email"
        placeholder="Email"
        registerOptions={{
          required: 'This is required.',
          maxLength: {
            value: 15,
            message: 'This phone number is too long.',
          },
        }}
      />,
      saveData,
    );
    const registeredInput = screen.getByPlaceholderText('Email');
    const submitButton = screen.getByTestId('testSubmitButton');

    // verify initial state
    expect(screen.queryByAltText('This is required.')).not.toBeInTheDocument();
    expect(
      screen.queryByAltText('This phone number is too long.'),
    ).not.toBeInTheDocument();

    // verify state after clicking submit button with empty inputs
    userEvent.click(submitButton);
    expect(await screen.findByText('This is required.')).toBeInTheDocument();
    expect(saveData).not.toHaveBeenCalled();

    //try submitting with an email that is not valid.
    userEvent.type(registeredInput, 'ThisIsAReallyLongEmailAddress@gmail.com');
    expect(
      await screen.findByText('This phone number is too long.'),
    ).toBeInTheDocument();
    userEvent.click(submitButton);
    await waitFor(() => {
      expect(saveData).not.toHaveBeenCalled();
    });

    //try an email that is valid.
    userEvent.clear(registeredInput);
    userEvent.type(registeredInput, 'ex@gmail.com');
    await waitFor(() => {
      expect(
        screen.queryByText('This phone number is too long.'),
      ).not.toBeInTheDocument();
    });

    // submit information
    userEvent.click(submitButton);
    await waitFor(() => {
      expect(saveData).toHaveBeenCalledWith({ email: 'ex@gmail.com' });
    });
  });
});
