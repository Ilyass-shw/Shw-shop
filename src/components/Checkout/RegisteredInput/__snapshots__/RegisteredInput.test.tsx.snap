// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisteredInput should match snapshot 1`] = `
<DocumentFragment>
  <form>
    <input
      autocapitalize="off"
      autocomplete="email"
      class="sc-bdvvaa gBkXga"
      name="email"
      placeholder="Email"
      spellcheck="false"
      type="email"
    />
    <p
      class="sc-gsDJrp eNrfob"
    />
    <button
      data-testid="testSubmitButton"
      type="submit"
    />
  </form>
</DocumentFragment>
`;
